### PR TITLE
conditionally defined var better error message

### PR DESCRIPTION
**Pull request by eellison**
i will do loops in a follow up after some other changes I am working on have landed

### CODE REVIEW

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/script/compiler.cpp [hunk 1]
```diff
@@ -1194,14 +1194,25 @@ struct to_ir {
     // ordered set, because we want deterministic graph output
     std::set<std::string> mutated_variables;
 
+    // var is only defined in one branch save error in case it's used later
+    auto err = [&](const std::string& v, const std::string& branch) {
+      ErrorReport error(stmt);
```
Comment (Krovatkin) [inline torch/csrc/jit/script/compiler.cpp]: this will get thrown when we try to do `getVar` to unify value types?

Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: No, this uses the same mechanism we use when we cannot unify the type of two values. We don't emit an error now, and cache the error msg if they try to use the var again later.  

https://github.com/pytorch/pytorch/blob/4d676d53a67e610689afbb34deabcac6ba8b8c50/torch/csrc/jit/script/compiler.cpp#L429

**Pull request closed without merging**